### PR TITLE
Refactor daily CI workflow

### PR DESCRIPTION
**Pull request by ydshieh**
# What does this PR do?

This is a re-work of  #29197. The main difference is: in this new work, we 
  - keep the definition of all jobs in a single workflow file (`self-scheduled.yml`)
    - this is changed to be a **re-usable** workflow
  - create a new workflow file (`self-scheduled-caller.yml`) as the caller to the above one (`self-scheduled.yml`)
    - it pass some inputs to the `self-scheduled.yml`

The reasons of the re-work:
  - **avoid too many workflow files** for a single CI. (As we also have `push-ci`, `amd-ci` x 2, `nightly-ci`, `past-ci` etc)

Once approved and merged, the same refactorization will be applied to other workflow files.

I have made sure it runs successfully as expected: https://github.com/huggingface/transformers/actions/runs/8524250043
(please don't look at the diagram - look the left side instead)

### CODE REVIEW

File: .github/workflows/self-scheduled-caller.yml [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+      - run_scheduled_ci*
+
+jobs:
+  model-ci:
+    name: Model CI
+    uses: ./.github/workflows/self-scheduled.yml
+    with:
+      job: run_tests_gpu
+      # See the comment for `ENV_NAME_FOR_CI_SLACK_REPORT_CHANNEL_ID` in `.github/workflows/slack-report.yml`.
+      env_name_for_slack_report_channel: CI_SLACK_CHANNEL_DUMMY_TESTS
```
Comment (ydshieh) [inline .github/workflows/self-scheduled-caller.yml:19]: Before merge: change to the correct channel (env.) names

File: .github/workflows/self-scheduled-caller.yml [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    secrets: inherit
+
+  torch-pipeline:
+    name: Torch pipeline CI
+    uses: ./.github/workflows/self-scheduled.yml
+    with:
+      job: run_pipelines_torch_gpu
+      # See the comment for `ENV_NAME_FOR_CI_SLACK_REPORT_CHANNEL_ID` in `.github/workflows/slack-report.yml`.
+      env_name_for_slack_report_channel: CI_SLACK_CHANNEL_DUMMY_TESTS
```
Comment (ydshieh) [inline .github/workflows/self-scheduled-caller.yml:28]: same as above, and for all occurrence below

File: .github/workflows/self-scheduled-caller.yml (hunk 1 above)
Comment (ydshieh) [inline .github/workflows/self-scheduled-caller.yml:19]: - the input `job` is used to control which job defined in `self-scheduled.yml` is to be run
  - (and used in  `notification_service.py` to control some reporting logic)
- the input `env_name_for_slack_report_channel` is used to control which channel to report for a specific job

File: utils/notification_service.py [hunk 3]
```diff
@@ -1056,6 +1074,24 @@ def prepare_reports(title, header, reports, to_truncate=True):
...
+        "run_all_tests_torch_cuda_extensions_gpu": "Torch CUDA extension tests",
+        "run_tests_quantization_torch_gpu": "Quantization tests",
+    }
+
+    # Remove some entries in `additional_files` if they are not concerned.
+    test_name = None
+    job_name = os.getenv("CI_TEST_JOB")
+    if job_name in job_to_test_map:
+        test_name = job_to_test_map[job_name]
+    additional_files = {k: v for k, v in additional_files.items() if k == test_name}
```
Comment (ArthurZucker) [inline utils/notification_service.py:1093]: were we testing additional things in some jobs?

File: .github/workflows/self-scheduled.yml [hunk 4]
```diff
@@ -396,52 +393,24 @@ jobs:
...
+    if: ${{ always() }}
+    uses: ./.github/workflows/slack-report.yml
+    with:
+      job: ${{ inputs.job }}
+      # This would be `skipped` if `setup` is skipped.
+      setup_status: ${{ needs.setup.result }}
+      env_name_for_slack_report_channel: ${{ inputs.env_name_for_slack_report_channel }}
+      # This would be an empty string if `setup` is skipped.
+      folder_slices: ${{ needs.setup.outputs.folder_slices }}
+    secrets: inherit
```
Comment (ArthurZucker) [inline .github/workflows/self-scheduled.yml:416]: very nice

File: .github/workflows/self-scheduled.yml [hunk 5]
```diff
@@ -348,18 +352,11 @@ jobs:
           path: /transformers/reports/${{ matrix.machine_type }}_tests_quantization_torch_gpu
 
   run_extract_warnings:
+    # Let's only do this for the job `run_tests_gpu` to simplify the (already complex) logic.
+    if: ${{ always() && inputs.job == 'run_tests_gpu' }}
```
Comment (ArthurZucker) [inline .github/workflows/self-scheduled.yml:356]: good call

File: .github/workflows/self-scheduled.yml
Comment (ArthurZucker) [inline .github/workflows/self-scheduled.yml:1]: Mostly the order of jobs is changed?

**Review by ArthurZucker — APPROVED** [review]
🔥 Let's not forget to update slack channel ids, but LGTM

File: utils/notification_service.py (hunk 3 above)
Comment (ydshieh) [reply utils/notification_service.py:1093]: No. That `additional` means the test jobs that are not the `modeling test jobs`: deepspeed, quantization, example, pipeline test jobs.

**Pull request merged**